File: com/as.tool/lua/script/flashloader.lua
-- /**
-- * AS - the open source Automotive Software on https://github.com/parai
-- *
-- * Copyright (C) 2015  AS <parai@foxmail.com>
-- *
-- * This source code is free software; you can redistribute it and/or modify it
-- * under the terms of the GNU General Public License version 2 as published by the
-- * Free Software Foundation; See <http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt>.
-- *
-- * This program is distributed in the hope that it will be useful, but
-- * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
-- * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
-- * for more details.
-- */
-- ===================== [ INCLUDE  ] ================================
require("dcm")
require("as")
require("s19")
require("math")
-- ===================== [ MACRO    ] ================================

-- ===================== [ LOCAL    ] ================================
local can_bus = 0
local dcm_chl = 0

local FLASH_WRITE_SIZE = 512
local FLASH_READ_SIZE  = 512
-- ===================== [ DATA     ] ================================
-- ===================== [ FUNCTION ] ================================
function enter_extend_session()
  ercd,res = dcm.transmit(dcm_chl,{0x10,0x03})
  
  if (false == ercd) then
    print("  >> enter extend session failed!")
  else
    print("  >> enter extend session ok!")
  end
  
  return ercd
  
end

function enter_program_session()
  ercd,res = dcm.transmit(dcm_chl,{0x10,0x02})
  
  if (false == ercd) then
    print("  >> enter program session failed!")
  else
    print("  >> enter program session ok!")
  end
  
  return ercd
  
end

function security_extds_access()
  -- level 1 
  ercd,res = dcm.transmit(dcm_chl,{0x27,0x01})
  
  if (false == ercd) then
    print("  >> security access request seed failed!")
  else
    print("  >> security access request seed ok!")
    seed = (res[3]<<24) + (res[4]<<16) + (res[5]<<8) +(res[6]<<0)
    key = seed ~ 0x78934673
    print(type(key),key)
    ercd,res = dcm.transmit(dcm_chl,{0x27,0x02,(key>>24)&0xFF,(key>>16)&0xFF,(key>>8)&0xFF,(key>>0)&0xFF})
    if (false == ercd) then
      print("  >> security access send key failed!")
    else
      print("  >> security access send key ok!")
    end
  end
  
  return ercd
  
end

function security_prgs_access()
  -- level 2
  ercd,res = dcm.transmit(dcm_chl,{0x27,0x03})
  
  if (false == ercd) then
    print("  >> security access request seed failed!")
  else
    print("  >> security access request seed ok!")
    seed = (res[3]<<24) + (res[4]<<16) + (res[5]<<8) +(res[6]<<0)
    key = seed ~ 0x94586792
    ercd,res = dcm.transmit(dcm_chl,{0x27,0x04,(key>>24)&0xFF,(key>>16)&0xFF,(key>>8)&0xFF,(key>>0)&0xFF})
    if (false == ercd) then
      print("  >> security access send key failed!")
    else
      print("  >> security access send key ok!")
    end
  end
  
  return ercd
  
end


function routine_erase_flash()
  ercd,res = dcm.transmit(dcm_chl,{0x31,0x01,0xFF,0x01,0x00,0x01,0x00,0x00,0x00,0x03,0x00,0x00,0xFF})
  -- start address = 0x00010000 = 64K 
  -- end   address = 0x00040000 = 256K, so size =  0x00030000
  -- identifier 0xFF
  
  if (false == ercd) then
    print("  >> routine erase flash failed!")
  else
    print("  >> routine erase flash ok!")
  end
  
  return ercd
  
end

function request_upload(addr,size,mem)
  data = {}
  data[1] = 0x35
  data[2] = 0x00 -- data format identifier
  data[3] = 0x44 -- address and length format
  data[4] = (addr>>24)&0xFF
  data[5] = (addr>>16)&0xFF
  data[6] = (addr>>8)&0xFF
  data[7] = (addr>>0)&0xFF
  data[8] = (size>>24)&0xFF
  data[9] = (size>>16)&0xFF
  data[10] = (size>>8)&0xFF
  data[11] = (size>>0)&0xFF
  data[12] = mem -- memory identifier
 
  ercd,res = dcm.transmit(dcm_chl,data)
  
  if (false == ercd) then
    print("  >> request upload failed!")
  else
    print("  >> request upload ok!")
  end
  
  return ercd
end

function request_transfer_exit()
  ercd,res = dcm.transmit(dcm_chl,{0x37})
  
  if (false == ercd) then
    print("  >> request_transfer_exit failed!")
  else
    print("  >> request_transfer_exit ok!")
  end
  
  return ercd
end

function request_download(addr,size,mem)
  data = {}
  data[1] = 0x34
  data[2] = 0x00 -- data format identifier
  data[3] = 0x44 -- address and length format
  data[4] = (addr>>24)&0xFF
  data[5] = (addr>>16)&0xFF
  data[6] = (addr>>8)&0xFF
  data[7] = (addr>>0)&0xFF
  data[8] = (size>>24)&0xFF
  data[9] = (size>>16)&0xFF
  data[10] = (size>>8)&0xFF
  data[11] = (size>>0)&0xFF
  data[12] = mem -- memory identifier
 
  ercd,res = dcm.transmit(dcm_chl,data)
  
  if (false == ercd) then
    print("  >> request download failed!")
  else
    print("  >> request download ok!")
  end
  
  return ercd
end

function request_transfer_exit()
  ercd,res = dcm.transmit(dcm_chl,{0x37})
  
  if (false == ercd) then
    print("  >> request_transfer_exit failed!")
  else
    print("  >> request_transfer_exit ok!")
  end
  
  return ercd
end

function download_one_record(addr,size,data,mem)
  ercd = request_download(addr,size,mem)
  
  -- download application
  blockSequenceCounter = 1
  left_size = size
  pos = 1
  ability = math.floor((4096-4)/FLASH_WRITE_SIZE) * FLASH_WRITE_SIZE

  while (left_size > 0) and (true== ercd) do
    req ={}
    req[1] = 0x36
    req[2] = blockSequenceCounter
    req[3] = 0
    req[4] = mem
    
    if (left_size > ability) then
      sz = ability
      left_size = left_size - ability
    else
      sz = math.floor((left_size+FLASH_WRITE_SIZE-1)/FLASH_WRITE_SIZE)*FLASH_WRITE_SIZE
      left_size = 0
    end
    for i=pos,size,1 do
      req[5+i-pos] = data[i]
      
      if((i-pos+1) >= sz) then
        break
      end
    end

    ercd,res = dcm.transmit(dcm_chl,req)
    
    pos = pos + sz

    blockSequenceCounter = (blockSequenceCounter + 1)&0xFF
 
  end
  
  if (true == ercd) then
    ercd = request_transfer_exit()
  end
  
  return ercd
end

function upload_one_record(addr,size,mem)
  ercd = request_upload(addr,size,mem)
  record = {}
  
  -- request upload application
  blockSequenceCounter = 1
  left_size = size
  pos = 0
  ability = math.floor((4096-4)/FLASH_READ_SIZE) * FLASH_READ_SIZE

  while (left_size > 0) and (true== ercd) do
    req ={}
    req[1] = 0x36
    req[2] = blockSequenceCounter
    req[3] = 0
    req[4] = mem

    ercd,res = dcm.transmit(dcm_chl,req)
    
    sz = rawlen(res) - 2
    
    if (left_size > sz) then
      left_size = left_size - sz
    else
      left_size = 0
    end
    
    for i=1,sz,1 do
      record[pos+i] = res[2+i]
    end
    
    pos = pos + sz

    blockSequenceCounter = (blockSequenceCounter + 1)&0xFF
 
  end
  
  if (true == ercd) then
    ercd = request_transfer_exit()
  end
  
  return ercd,record
end

function download_flash_driver()
  srecord = s19.open("D:/repository/as/release/asboot/out/stm32f107vc-flsdrv.s19")
  
  if( nil == srecord ) then
    print("  >> invalid flash driver srecord file!")
    return false
  end
  -- flash driver mapped to address 0
  secnbr = rawlen(srecord)
  for i=1,secnbr,1 do
    ss = srecord[i]
    addr =  ss['addr']-srecord[1]['addr']
    ercd =  download_one_record(addr,ss['size'],ss['data'],0xFD)
    if (false == ercd) then
      break
    end
  end
 
  if (false == ercd) then
    print("  >> download flash driver failed!")
  else
    print("  >> download flash driver ok!")
  end
  
  return ercd
end

function fl_compare(s1,s2)

  ercd = true
  length = rawlen(s1)
  for i=1,length,1 do
    if s1[i] ~= s2[i] then
      ercd = false
      break
    end
  end
  
  return ercd
end

function check_flash_driver()
  srecord = s19.open("D:/repository/as/release/asboot/out/stm32f107vc-flsdrv.s19")
  
  if( nil == srecord ) then
    print("  >> invalid flash driver srecord file!")
    return false
  end
  -- flash driver mapped to address 0
  secnbr = rawlen(srecord)
  for i=1,secnbr,1 do
    ss = srecord[i]
    addr =  ss['addr']-srecord[1]['addr']
    ercd,record =  upload_one_record(addr,ss['size'],0xFD)
    if (false == ercd) then
      break
    else
      ercd = fl_compare(srecord[i]['data'],record)
      if (false == ercd) then
        break
      end
    end
  end
 
  if (false == ercd) then
    print("  >> check flash driver failed!")
  else
    print("  >> check flash driver ok!")
  end
  
  return ercd
end

function download_application()

  srecord = s19.open("D:/repository/as/release/ascore/out/stm32f107vc.s19")
  
  if( nil == srecord ) then
    print("  >> invalid application srecord file!")
    return false
  end
  
  secnbr = rawlen(srecord)
  for i=1,secnbr,1 do
    ss = srecord[i]
    addr =  ss['addr']
    ercd =  download_one_record(addr,ss['size'],ss['data'],0xFF)
    if (false == ercd) then
      break
    end
  end
 
  if (false == ercd) then
    print("  >> download application failed!")
  else
    print("  >> download application ok!")
  end
  
  return ercd
end

function check_application()
  srecord = s19.open("D:/repository/as/release/ascore/out/stm32f107vc.s19")
  
  if( nil == srecord ) then
    print("  >> invalid application srecord file!")
    return false
  end
  secnbr = rawlen(srecord)
  for i=1,secnbr,1 do
    ss = srecord[i]
    addr =  ss['addr']
    ercd,record =  upload_one_record(addr,ss['size'],0xFF)
    if (false == ercd) then
      break
    else
      ercd = fl_compare(srecord[i]['data'],record)
      if (false == ercd) then
        break
      end
    end
  end
 
  if (false == ercd) then
    print("  >> check application failed!")
  else
    print("  >> check application ok!")
  end
  
  return ercd
end

function routine_test_jump_to_application()
  ercd,res = dcm.transmit(dcm_chl,{0x31,0x01,0xFF,0x03})
  
  if (false == ercd) then
    print("  >> routine test jump to application failed!")
  else
    print("  >> routine test jump to application ok!")
  end
  
  return ercd
  
end

operation_list = {enter_extend_session, security_extds_access,
                  enter_program_session,security_prgs_access,
                  download_flash_driver,check_flash_driver,
                  routine_erase_flash, download_application,check_application,
				          routine_test_jump_to_application
}

function main()
  data = {}
  as.can_open(can_bus,"rpmsg",0,1000000)
  -- as.can_open(can_bus,"serial",3,115200)	-- COM4
  dcm.init(dcm_chl,can_bus,0x732,0x731)
  
  
  for i=1,rawlen(operation_list),1 do
    ercd = operation_list[i]()
    if false == ercd then
      break
    end
  end
  
end

main()

File: com/as.tool/lua/script/flashloader.py
from email._header_value_parser import Address
__lic__ = '''
/**
 * AS - the open source Automotive Software on https://github.com/parai
 *
 * Copyright (C) 2015  AS <parai@foxmail.com>
 *
 * This source code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation; See <http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt>.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */
 '''
 
from .dcm import *
from .s19 import *


from PyQt5 import QtCore, QtGui
from PyQt5.QtGui import *
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *

import os

__all__ = ['UIFlashloader']

class AsFlashloader(QThread):
    infor = QtCore.pyqtSignal(str)
    progress = QtCore.pyqtSignal(int)
    def __init__(self,parent=None):
        super(QThread, self).__init__(parent)
        self.steps = [ self.enter_extend_session, self.security_extds_access,
                  self.enter_program_session,self.security_prgs_access,
                  self.download_flash_driver,
                  self.routine_erase_flash, self.download_application ]
        self.enable = []
        for s in self.steps:
            self.enable.append(True)
        self.dcm = dcm(0,0x732,0x731)
        self.progress_value = 0
        self.app = None
        self.flsdrv = None

    def setTarget(self,app,flsdrv=None):
        self.app = app
        self.flsdrv = flsdrv

    def GetSteps(self):
        ss = []
        for s in self.steps:
            ss.append(s.__name__.replace('_',' '))
        return ss
    
    def SetEnable(self,step,enable):
        for id,s in enumerate(self.steps):
            if(step == s.__name__.replace('_',' ')):
                self.enable[id] = enable

    def transmit(self,req,exp):
        ercd,res = self.dcm.transmit(req)
        if(ercd == True):
            if(len(res)>=len(exp)):
                for i in range(len(exp)):
                    if((res[i]!=exp[i]) and (exp[i]!=-1)):
                        ercd = False
                        break
            else:
                ercd = False
        if(ercd == True): 
            self.infor.emit('  success')
            self.progress.emit(self.progress_value)
        else:
            self.infor.emit('  failed')
        return ercd,res
    def enter_extend_session(self):
        return self.transmit([0x10,0x03], [0x50,0x03])
    def security_extds_access(self):
        ercd,res = self.transmit([0x27,0x01], [0x67,0x01,-1,-1,-1,-1])
        if(ercd):
            seed = (res[2]<<24) + (res[3]<<16) + (res[4]<<8) +(res[5]<<0)
            key = (seed^0x78934673)
            self.infor.emit(' send key %X from seed %X'%(key,seed))
            ercd,res = self.transmit([0x27,0x02,(key>>24)&0xFF,(key>>16)&0xFF,(key>>8)&0xFF,(key>>0)&0xFF],[0x67,0x02])
        return ercd,res
    def enter_program_session(self):
        return self.transmit([0x10,0x02], [0x50,0x02])
    def security_prgs_access(self):
        ercd,res = self.transmit([0x27,0x03], [0x67,0x03,-1,-1,-1,-1])
        if(ercd):
            seed = (res[2]<<24) + (res[3]<<16) + (res[4]<<8) +(res[5]<<0)
            key = (seed^0x94586792)
            self.infor.emit(' send key %X from seed %X'%(key,seed))
            ercd,res = self.transmit([0x27,0x04,(key>>24)&0xFF,(key>>16)&0xFF,(key>>8)&0xFF,(key>>0)&0xFF],[0x67,0x04])
        return ercd,res
    def request_download(self,address,size,identifier):
        self.infor.emit(' request download')
        return self.transmit([0x34,0x00,0x44,     \
                            (address>>24)&0xFF,(address>>16)&0xFF,(address>>8)&0xFF,(address>>0)&0xFF,   \
                            (size>>24)&0xFF,(size>>16)&0xFF,(size>>8)&0xFF,(size>>0)&0xFF,  \
                            identifier],[0x74])

    def request_upload(self,address,size,identifier):
        self.infor.emit(' request upload')
        return self.transmit([0x35,0x00,0x44,     \
                            (address>>24)&0xFF,(address>>16)&0xFF,(address>>8)&0xFF,(address>>0)&0xFF,   \
                            (size>>24)&0xFF,(size>>16)&0xFF,(size>>8)&0xFF,(size>>0)&0xFF,  \
                            identifier],[0x75])
    
    def request_transfer_exit(self):
        self.infor.emit(' request transfer exit')
        return self.transmit([0x37],[0x77])
    
    def download_one_section(self,address,size,data,identifier):
        FLASH_WRITE_SIZE = 512
        blockSequenceCounter = 1
        left_size = size
        pos = 0
        ability = int(((4096-4)/FLASH_WRITE_SIZE)) * FLASH_WRITE_SIZE
        ercd,res = self.request_download(address,size,identifier)
        if(ercd == False):return ercd,res
        while(left_size>0 and ercd==True):
            req = [0x36,blockSequenceCounter,0,identifier]
            if(left_size > ability):
                sz = ability
            else:
                sz = int((left_size+FLASH_WRITE_SIZE-1)/FLASH_WRITE_SIZE)*FLASH_WRITE_SIZE
                left_size = 0
            for i in range(sz):
                if((pos+i)<size):
                    req.append(data[pos+i])
                else:
                    req.append(0xFF)
            self.infor.emit(' transfer block %s'%(blockSequenceCounter))
            ercd,res = self.transmit(req,[0x76,blockSequenceCounter])
            if(ercd == False):return ercd,res
            blockSequenceCounter = (blockSequenceCounter + 1)&0xFF
            pos += sz
        ercd,res = self.request_transfer_exit()
        if(ercd == False):return ercd,res
        return ercd,res

    def upload_one_section(self,address,size,identifier):
        FLASH_READ_SIZE = 512
        blockSequenceCounter = 1
        left_size = size
        ability = int(((4096-4)/FLASH_READ_SIZE)) * FLASH_READ_SIZE
        ercd,res = self.request_upload(address,size,identifier)
        if(ercd == False):return ercd,res,None
        data = []
        while(left_size>0 and ercd==True):
            req = [0x36,blockSequenceCounter,0,identifier]
            self.infor.emit(' transfer block %s'%(blockSequenceCounter))
            ercd,res = self.transmit(req,[0x76,blockSequenceCounter])
            if(ercd == False):return ercd,res,None
            blockSequenceCounter = (blockSequenceCounter + 1)&0xFF
            sz = len(res)-2
            if (left_size > sz):
                left_size = left_size - sz
            else:
                left_size = 0
            for b in res[2:]:
                data.append(b)
        ercd,res = self.request_transfer_exit()
        if(ercd == False):return ercd,res,None
        return ercd,res,data
    
    def compare(self,d1,d2):
        for i,b in enumerate(d1):
            if(b!=d2[i]):return False
        return True

    def download_flash_driver(self):
        flsdrv = s19(self.flsdrv)
        ary = flsdrv.getData()
        for ss in ary:
            ercd,res = self.download_one_section(ss['address']-ary[0]['address'],ss['size'],ss['data'],0xFD)
            if(ercd == False):return ercd,res
        
        for ss in ary:
            ercd,res,up = self.upload_one_section(ss['address']-ary[0]['address'],ss['size'],0xFD)
            if(ercd and self.compare(ss['data'], up)):
                self.infor.emit('  check flash driver pass!')
            else:
                self.infor.emit('  check flash driver failed!')
                return False,res
        return ercd,res
    
    def routine_erase_flash(self):
        return self.transmit([0x31,0x01,0xFF,0x01,0x00,0x01,0x00,0x00,0x00,0x03,0x00,0x00,0xFF],[0x71,0x01,0xFF,0x01])
    
    def download_application(self):
        app = s19(self.app)
        ary = app.getData()
        for ss in ary:
            ercd,res = self.download_one_section(ss['address'],ss['size'],ss['data'],0xFF)
            if(ercd == False):return ercd,res
        
        for ss in ary:
            ercd,res,up = self.upload_one_section(ss['address'],ss['size'],0xFF)
            if(ercd and self.compare(ss['data'], up)):
                self.infor.emit('  check application pass!')
            else:
                self.infor.emit('  check application failed!')
                return False,res
        return ercd,res
    
    def run(self):
        self.infor.emit("starting ... ")
        self.progress_value = 0
        for id,s in enumerate(self.steps):
            if(self.enable[id] == True):
                self.progress_value += 50/(len(self.steps))
                self.infor.emit('>> '+s.__name__.replace('_',' '))
                ercd,res = s()
                if(ercd == False):
                    self.infor.emit("\n\n  >> boot failed <<\n\n")
                    return
        self.progress.emit(100)

class AsStepEnable(QCheckBox):
    enableChanged=QtCore.pyqtSignal(str,bool)
    def __init__(self,text,parent=None):
        super(QCheckBox, self).__init__(text,parent)
        self.stateChanged.connect(self.on_stateChanged)
    def on_stateChanged(self,state):
        self.enableChanged.emit(self.text(),state)
        
class UIFlashloader(QWidget):
    def __init__(self, parent=None):
        super(QWidget, self).__init__(parent)
        
        self.loader = AsFlashloader()
        self.loader.infor.connect(self.on_loader_infor)
        self.loader.progress.connect(self.on_loader_progress)
        
        vbox = QVBoxLayout()
        
        grid = QGridLayout()
        grid.addWidget(QLabel('Application'),0,0)
        self.leApplication = QLineEdit()
        grid.addWidget(self.leApplication,0,1)
        self.btnOpenApp = QPushButton('...')
        grid.addWidget(self.btnOpenApp,0,2)
        
        grid.addWidget(QLabel('Flash Driver'),1,0)
        self.leFlsDrv = QLineEdit()
        grid.addWidget(self.leFlsDrv,1,1)
        self.btnOpenFlsDrv = QPushButton('...')
        grid.addWidget(self.btnOpenFlsDrv,1,2)

        grid.addWidget(QLabel('Progress'),2,0)
        self.pgbProgress = QProgressBar()
        self.pgbProgress.setRange(0,100)
        grid.addWidget(self.pgbProgress,2,1)
        self.btnStart=QPushButton('Start')
        grid.addWidget(self.btnStart,2,2)
        vbox.addLayout(grid)
        
        hbox = QHBoxLayout()
        vbox2 = QVBoxLayout()
        for s in self.loader.GetSteps():
            cbxEnable = AsStepEnable(s)
            cbxEnable.setChecked(True)
            cbxEnable.enableChanged.connect(self.on_enableChanged)
            vbox2.addWidget(cbxEnable)
        hbox.addLayout(vbox2)
        self.leinfor = QTextEdit()
        self.leinfor.setReadOnly(True)
        hbox.addWidget(self.leinfor)
        
        vbox.addLayout(hbox)
        
        self.setLayout(vbox)
        
        self.btnOpenApp.clicked.connect(self.on_btnOpenApp_clicked)
        self.btnOpenFlsDrv.clicked.connect(self.on_btnOpenFlsDrv_clicked)
        self.btnStart.clicked.connect(self.on_btnStart_clicked)
        
        self.app=''
        self.flsdrv=''
        
        default_app = '/home/parai/workspace/as/release/ascore/out/stm32f107vc.s19'
        default_flsdrv = '/home/parai/workspace/as/release/asboot/out/stm32f107vc-flsdrv.s19'
        if(os.path.exists(default_app)):
            self.app = default_app
            self.leApplication.setText(default_app)
        if(os.path.exists(default_flsdrv)):
            self.flsdrv = default_flsdrv
            self.leFlsDrv.setText(default_flsdrv)

    def on_enableChanged(self,step,enable):
        self.loader.SetEnable(step, enable)

    def on_loader_infor(self,text):
        self.leinfor.append(text)
    
    def on_loader_progress(self,prg):
        self.pgbProgress.setValue(prg)

    def on_btnOpenApp_clicked(self):
        rv = QFileDialog.getOpenFileName(None,'application file', '','application (*.s19 *.bin)')
        self.app = rv[0]
        self.leApplication.setText(self.app)

    def on_btnOpenFlsDrv_clicked(self):
        rv = QFileDialog.getOpenFileName(None,'flash driver file', '','flash driver (*.s19 *.bin)')
        self.flsdrv = rv[0]
        self.leFlsDrv.setText(self.flsdrv)


    def on_btnStart_clicked(self):
        if(os.path.exists(self.app)):
            self.pgbProgress.setValue(1)
            self.loader.setTarget(self.app, self.flsdrv)
            self.loader.start()
        else:
            QMessageBox.information(self, 'Tips', 'Please load a valid application first!')
        